Burn Perimeter, FBP Outputs, and Resampling Record Keeping Update

shreeramsenthi opened the pull request:
Please see the accompanying beta release for a summary of changes and a release file to test against:

https://github.com/BurnP3/BurnP3Plus/releases/tag/2.4.0

Please also note you will need SyncroSim v3.1.9 for this release.

shreeramsenthi committed:
Major updates to burn perimeters and FBP outputs

DB and XML update for consolidating FBP outputs and perimeter changes

Update output option validation for final or daily burn perimeters

Consolidate burn perimeter geopackage in summary transformer

Fix bug with trying to save empty datasheets

Clean up missing value handling in spatial output options

Update calculation and organization of FBP summaries

shreeramsenthi committed:
Temporarily hide percentile FBP summaries

shreeramsenthi committed:
Update to resampling record keeping

shreeramsenthi committed:
Clean up merge and package versions

katieb1 commented:
I tried running a FireSTARR model with 500 iterations (that generated 34,850 fires) and selecting the FBP output for median and max fire intensity, but ran into what looks like a memory error when applying the maximum function to the raster stack. Also BurnP3+ can't seem to find the burn perimeters GeoPackage - I didn't select the burn perimeters output option, so maybe that's why? But then I don't think BurnP3+ should be looking for the GeoPackage if it doesn't create it in the first place.

See error message (full run log also attached):
```
[Failure] The external program failed.
Program name was: C:\Users\Administrator\AppData\Local\SyncroSim Studio\Conda\envs\burnP3Plus-2\burnP3Plus\Scripts\Rscript.exe
Arguments were: "C:\Users\Administrator\AppData\Local\SyncroSim\Packages\burnP3Plus\2.4.0\burnProbability.R"
Exit code was: 1
More information:
Warning message:
In Sys.timezone() : unable to identify current timezone 'C':
please set environment variable 'TZ'
Saved
Saved
Saved
Warning message:
In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="C:/projects/R910/GreenShield.ssim.temp/summary/burn-perimeters.gpkg": The system cannot find the file specified
Error in x@ptr$summary(fun, na.rm, opt) : std::bad_alloc
Calls: max -> max -> .summarize -> <Anonymous> -> .External
Execution halted
```

[2025-03-13_BP3_test_run.txt](https://github.com/user-attachments/files/20199207/2025-03-13_BP3_test_run.txt)

I should also mention that I'm using the c5a.8xlarge AWS instance type, which has 32 vCPUs and 64 GB of ram. Is it possible to memory optimize the FBP output summarizing functions? I imagine most people will be doing much larger runs than me with similar specs. Possibly the raster calculations could be done in subsets of the raster stacks (i.e., divide the raster stack into stacks of 100, apply function to reduce the stack to a single raster, then apply functions to stack of 100 reduced rasters, etc. until a single summarized raster is produced).

katieb1 commented:
Update - got rid of the extra ignitions so only 6212 fires burned and I'm still getting an out of memory error when summarizing across the entire raster stack.

katieb1 commented:
Sorry for the multiple comments, but I had another thought about my above potential "solution". It won't work for medians. BUT, it looks like `terra` can process rasters in chunks using the `terra::app()` function. It might be worth just trying to swapp out the following code as a first pass:

`fbpSummaryMap <- max(fbpStack, na.rm = T)`

for 

`fbpSummaryMap <- app(fbpStack, max, na.rm = TRUE)`

shreeramsenthi committed:
Ignore warnings when normalizing paths

shreeramsenthi commented:
Thanks for testing things! Let's start with the easy one, I forgot to add an argument (`mustWork = F`) when normalizing paths for the geopackage. Should be fixed now!

For the memory things, this line: https://github.com/BurnP3/BurnP3Plus/blob/5241c3ba603761b624ecf083097d015c71cec98b/src/burnProbability.R#L38 should limit `terra` function memory use to 2GB which seemed to be working on my end but I guess not for you. Any idea if that might be a `terra` version / conda issue? If so I can back those out for DIY memory-safety. I'm not totally sure how to do that for the `median` function, but worst case we can hide that temporarily like the percentile functions.

I did try playing around with the `app` function, particularly when I was trying to implement memory-safe percentile summaries, but it was a bit slower in my testing and (I think) reads the same terra options variable for memory limits. Does using `app` make things memory safe on your instance? 

Edit: Here's a branch with the `app` changes to test with: https://github.com/BurnP3/BurnP3Plus/tree/testing-app-memory-safety

shreeramsenthi committed:
Fix bug with parsing spatial output options

shreeramsenthi committed:
Bug fixes to FBP output option parsing

shreeramsenthi committed:
Fix bug with plotting raw FBP outputs after summary

shreeramsenthi committed:
Fix memory safety for older versions of terra, enable percentile FBP summaries

shreeramsenthi commented:
Added some more bug fixes based on conversations with Tom, Katie, and Weiwei, added percentile FBP summaries back in, and added a fix for memory safety in older versions of terra (specifically the latest one available in Conda for Windows).

katieb1 committed:
Update version number

BadgerOnABike commented:
How does this factor with the move to arrow?

shreeramsenthi commented:
The branch with the arrow updates is built on top of this branch. Once we're happy with that we can probably merge the latest tag directly into dev instead of this PR.

shreeramsenthi commented:
Superseded by https://github.com/BurnP3/BurnP3Plus/pull/95 .
